### PR TITLE
Add pulsar-client-cpp

### PR DESCRIPTION
**Pull request by blackheaven**
###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

I'm not actually the original author ([see](https://github.com/CorbanR/nixpkgs/issues/2#issuecomment-974553064))

### CODE REVIEW

File: maintainers/maintainer-list.nix [hunk 1]
```diff
@@ -4181,6 +4181,12 @@
     githubId = 37017396;
     name = "gbtb";
   };
+  gdifolco = {
```
Comment (SuperSandro2000) [inline maintainers/maintainer-list.nix:4184]: Please put this into a separate commit with the message ``maintainers: add NAME`` and sort it alphabetically.

File: pkgs/development/libraries/pulsar-client-cpp/default.nix [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+        ${if stdenv.isDarwin
+        then "$CC++ test.cc -L $out/lib -I $out/include -lpulsar -o test"
+        else "g++ test.cc -L $out/lib -I $out/include -lpulsar -o test" }
+    '';
+
+    meta = with lib; {
+      homepage = "https://pulsar.apache.org/docs/en/client-libraries-cpp";
+      description = "Apache Pulsar C++ library";
+
+      platforms = [ "x86_64-darwin" "x86_64-linux" "aarch64-darwin"];
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/pulsar-client-cpp/default.nix:76]:
```suggestion
      description = "Apache Pulsar C++ library";
      platforms = [ "x86_64-darwin" "x86_64-linux" "aarch64-darwin"];
```

File: pkgs/development/libraries/pulsar-client-cpp/default.nix [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+    echo ${lib.escapeShellArg ''
+      #include <pulsar/Client.h>
+      int main (int argc, char **argv) {
+        pulsar::Client client("pulsar://localhost:6650");
+        return 0;
+      }
+    ''} > test.cc
+        ${if stdenv.isDarwin
+        then "$CC++ test.cc -L $out/lib -I $out/include -lpulsar -o test"
+        else "g++ test.cc -L $out/lib -I $out/include -lpulsar -o test" }
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/pulsar-client-cpp/default.nix:69]: We can't just set CC to g on darwin?

File: pkgs/development/libraries/pulsar-client-cpp/default.nix [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+    cmakeFlags = [
+      "-DBUILD_TESTS=OFF"
+      "-DBUILD_PYTHON_WRAPPER=OFF"
+      "-DClangTools_PATH=${clang-tools}/bin"
+    ];
+
+    patches = [ ./reader_configuration_includes.diff ];
+    enableParallelBuilding = true;
+
+  doInstallCheck = true;
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/pulsar-client-cpp/default.nix:58]: This is formatted inconsistently. nixpkgs-fmt can help you with formatting.

File: pkgs/development/libraries/pulsar-client-cpp/default.nix [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+    # since we cant expand $out in cmakeFlags
+    preConfigure = ''cmakeFlags="$cmakeFlags -DCMAKE_INSTALL_LIBDIR=$out/lib"'';
+
+    cmakeFlags = [
+      "-DBUILD_TESTS=OFF"
+      "-DBUILD_PYTHON_WRAPPER=OFF"
+      "-DClangTools_PATH=${clang-tools}/bin"
+    ];
+
+    patches = [ ./reader_configuration_includes.diff ];
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/pulsar-client-cpp/default.nix:55]: Is there an upstream patch we could fetch instead?

File: pkgs/development/libraries/pulsar-client-cpp/default.nix [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+    # python37 used in cmake script to calculate some values
+    # clang-tools needed for clang-format etc
+    nativeBuildInputs = [ cmake python37 pkgconfig ]
+    ++ lib.optional stdenv.isDarwin [ clang clang-tools ]
+    ++ lib.optional stdenv.isLinux [ gcc clang-tools ];
+
+    buildInputs = [ boost17x jsoncpp log4cxx openssl protobuf snappy zstd curl zlib ];
+
+    # since we cant expand $out in cmakeFlags
+    preConfigure = ''cmakeFlags="$cmakeFlags -DCMAKE_INSTALL_LIBDIR=$out/lib"'';
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/pulsar-client-cpp/default.nix:47]: We can with ``${placeholder "out"}``.

File: pkgs/development/libraries/pulsar-client-cpp/default.nix [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+      url = "https://archive.apache.org/dist/pulsar/pulsar-${version}/apache-pulsar-${version}-src.tar.gz";
+    };
+
+    sourceRoot = "apache-pulsar-${version}-src/pulsar-client-cpp";
+
+    # python37 used in cmake script to calculate some values
+    # clang-tools needed for clang-format etc
+    nativeBuildInputs = [ cmake python37 pkgconfig ]
+    ++ lib.optional stdenv.isDarwin [ clang clang-tools ]
+    ++ lib.optional stdenv.isLinux [ gcc clang-tools ];
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/pulsar-client-cpp/default.nix:42]:
```suggestion
    ++ lib.optionals stdenv.isDarwin [ clang clang-tools ]
    ++ lib.optionals stdenv.isLinux [ gcc clang-tools ];
```

File: pkgs/development/libraries/pulsar-client-cpp/default.nix [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+
+    src = fetchurl {
+      inherit sha512;
+      url = "https://archive.apache.org/dist/pulsar/pulsar-${version}/apache-pulsar-${version}-src.tar.gz";
+    };
+
+    sourceRoot = "apache-pulsar-${version}-src/pulsar-client-cpp";
+
+    # python37 used in cmake script to calculate some values
+    # clang-tools needed for clang-format etc
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/pulsar-client-cpp/default.nix:39]: Do we really need clang-format? Sounds like they are used in some sort of lint CI which we can happily skip.

Can we use python39 through python3? If so please patch the python37 out so that we don't need to worry about that.

File: pkgs/development/libraries/pulsar-client-cpp/default.nix [hunk 9]
```diff
@@ -0,0 +1,80 @@
+{ stdenv
+, lib
+, pkgs
+, fetchurl
+, version ? "2.8.1"
+, sha512 ? "23i95k6gs5lh6gw6v6hybv6z8rfzc3mfgj2agb6iicad9ia2sf46wr3rqzi9bdhp0c990vcvn72bkz13i0hnh1hj5gnwqin339a2g0x"
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/pulsar-client-cpp/default.nix:6]: Please inline this.

File: pkgs/development/libraries/pulsar-client-cpp/default.nix (hunk 5 above)
Comment (blackheaven) [reply pkgs/development/libraries/pulsar-client-cpp/default.nix:55]: I actually did a PR: https://github.com/apache/pulsar/pull/12966

But, I'm not sure how to integrate it.

File: pkgs/development/libraries/pulsar-client-cpp/default.nix (hunk 3 above)
Comment (blackheaven) [reply pkgs/development/libraries/pulsar-client-cpp/default.nix:69]: If I don't, it looks for `gcc++`, which is not found

**Comment by CorbanR** [comment]
This should already be done! https://github.com/NixOS/nixpkgs/commit/d59d843f217ccbeea57938be1eca3a15bdd71cb9

**Comment by blackheaven** [comment]
Thanks
